include/swift-c/DependencyScan/DependencyScan.h

```diff
@@ -0,0 +1,214 @@
+//===--- DependencyScan.h - C API for Swift Dependency Scanning ---*- C -*-===//
+//
+// This source file is part of the Swift.org open source project
+//
+// Copyright (c) 2014 - 2020 Apple Inc. and the Swift project authors
+// Licensed under Apache License v2.0 with Runtime Library Exception
+//
+// See https://swift.org/LICENSE.txt for license information
+// See https://swift.org/CONTRIBUTORS.txt for the list of Swift project authors
+//
+//===----------------------------------------------------------------------===//
+//
+// This C API is primarily intended to serve as the Swift Driver's
+// dependency scanning facility (https://github.com/apple/swift-driver).
+//
+//===----------------------------------------------------------------------===//
+
+#ifndef SWIFT_C_DEPENDENCY_SCAN_H
+#define SWIFT_C_DEPENDENCY_SCAN_H
+
+#include "DependencyScanMacros.h"
+#include "DSString.h"
+#include <stdbool.h>
+#include <stddef.h>
+#include <stdint.h>
+
+/// The version constants for the SwiftDependencyScan C API.
+/// DEPSCAN_VERSION_MINOR should increase when there are API additions.
+/// DEPSCAN_VERSION_MAJOR is intended for "major" source/ABI breaking changes.
+#define DEPSCAN_VERSION_MAJOR 0
+#define DEPSCAN_VERSION_MINOR 1
+
+DEPSCAN_BEGIN_DECLS
+
+//=== Scanner Data Types --------------------------------------------------===//
+
+typedef enum {
+  DEPSCAN_DEPENDENCY_INFO_SWIFT_TEXTUAL = 0,
+  DEPSCAN_DEPENDENCY_INFO_SWIFT_BINARY = 1,
+  DEPSCAN_DEPENDENCY_INFO_SWIFT_PLACEHOLDER = 2,
+  DEPSCAN_DEPENDENCY_INFO_CLANG = 3
+} depscan_dependency_info_kind_t;
+
+/// Swift modules to be built from a module interface, may have a bridging
+/// header.
+typedef struct {
+  /// The module interface from which this module was built, if any.
+  ds_string_t module_interface_path;
+
+  /// The paths of potentially ready-to-use compiled modules for the interface.
+  ds_string_set_t* compiled_module_candidates;
+
+  /// The bridging header, if any.
+  ds_string_t bridging_header_path;
+
+  /// The source files referenced by the bridging header.
+  ds_string_set_t* bridging_source_files;
+
+  /// (Clang) modules on which the bridging header depends.
+  ds_string_set_t* bridging_module_dependencies;
+
+  /// Options to the compile command required to build this module interface
+  ds_string_set_t* command_line;
+
+  /// To build a PCM to be used by this Swift module, we need to append these
+  /// arguments to the generic PCM build arguments reported from the dependency
+  /// graph.
+  ds_string_set_t* extra_pcm_args;
+
+  /// The hash value that will be used for the generated module
+  ds_string_t context_hash;
+
+  /// A flag to indicate whether or not this module is a framework.
+  bool is_framework;
+} depscan_swift_textual_details_t;
```

Comment:
Shouldn't we treat `depscan_swift_textual_details_t` as an opaque type and add C APIs to retrieve each field? This allows us to modify the actual contents in the details. I guess this also applies to all other `details` types in this header.

Reply:
Yes, will switch to this, thank you.

Reply:
I made this change. There is now an added file:
`include/swift/DependencyScan/DependencyScanImpl.h`, which contains the implementation of these data structures which we are free to modify, while the public C interface uses an opaque container for details with APIs to query fields. 
The implementation of the C API has been factored out into `lib/DependencyScan/DependencyScanImpl.cpp`.